crates/red_knot_python_semantic/src/semantic_index/visibility_constraints.rs

```diff
@@ -553,37 +554,99 @@ impl VisibilityConstraints {
         }
     }
 
+    fn analyze_single_pattern_predicate_kind<'db>(
+        db: &'db dyn Db,
+        predicate_kind: &PatternPredicateKind<'db>,
+        subject: Expression<'db>,
+    ) -> Truthiness {
+        match predicate_kind {
+            PatternPredicateKind::Value(value) => {
+                let subject_ty = infer_expression_type(db, subject);
+                let value_ty = infer_expression_type(db, *value);
+
+                if subject_ty.is_single_valued(db) {
+                    Truthiness::from(subject_ty.is_equivalent_to(db, value_ty))
+                } else {
+                    Truthiness::Ambiguous
+                }
+            }
+            PatternPredicateKind::Singleton(singleton) => {
+                let subject_ty = infer_expression_type(db, subject);
+
+                let singleton_ty = match singleton {
+                    ruff_python_ast::Singleton::None => Type::none(db),
+                    ruff_python_ast::Singleton::True => Type::BooleanLiteral(true),
+                    ruff_python_ast::Singleton::False => Type::BooleanLiteral(false),
+                };
+
+                if subject_ty.is_single_valued(db) {
+                    Truthiness::from(subject_ty.is_equivalent_to(db, singleton_ty))
```

Comment:
it should be always true here that the type of `singleton_ty` is recognised as a singleton type by red-knot. That means that we can skip checking whether `subject_ty` is single-valued and go straight to the equivalence check: if `subject_ty` is equivalent to `singleton_ty`, that means that `subject_ty` is a singleton type, which means it is also a single-valued type. (The set of singleton types is a subset of the set of single-valued types; all singleton types are single-valued.)

```suggestion
                debug_assert!(singleton_ty.is_singleton(db));

                if subject_ty.is_equivalent_to(db, singleton_ty) {
                    Truthiness::AlwaysTrue
```